### DISCUSSION TITLE
Setting MSI features using bootstrapper

### DISCUSSION THREAD

**Discussion by amyspark** [wixtoolset/issues#8843] [Questions] [answered]
Hi all,

Is there a way to set a MSI's features, when **not** using [WixInternalUIBootstrapperApplication (Bal extension)](https://wixtoolset.org/docs/schema/bal/wixinternaluibootstrapperapplication/)?

I've tried an example based on the layout [here](https://github.com/wixtoolset/wix/blob/f440fb317c630e7bc6d4ee4d657a200654e2f876/src/test/burn/TestData/LayoutTests/BundleA/Bundle.wxs#L26C1-L26C90) but the underlying MSI package always assumes a Typical instead of a Complete or Custom install. According to the options of MsiPackage,

> DisplayInternalUICondition (String, from extension namespace http://wixtoolset.org/schemas/v4/wxs/bal) : Specifies whether the bundle will show the UI authored into the msi package. If not specified or the condition evaluates to false, **all information is routed to the bootstrapper application to provide a unified installation experience**. Otherwise, the UI authored into the msi package will be displayed on top of any bootstrapper application UI. WixStdBA doesn't support EmbeddedUI.

but I can't seem to achieve the bit marked in bold.

**Comment by wmanning** [comment]
Use the <MsiProperty> element under <MsiPackage> to pass in properties. 
So pass in a property listing the features you want and then in the MSI package copy the value of that property to ADDLOCAL.  I think that is the way you would do it.

Or maybe something like this: https://stackoverflow.com/a/70429470/623561

IMHO the "modern" way to do features if you're using burn (aka bundle) is to have a MSI per feature. Then create a bundle BA to control which MSI packages get installed.

**Reply by amyspark** [reply]
In short, the only way to do this is to use a custom BA? (Whether to select the individual MSIs or to pass ADDLOCAL)

**Reply by barnson** [reply]
WixStdBA supports install conditions and passing properties to individual MsiPackages.

**Reply by amyspark** [reply]
Yes, but what we're looking for is an UI to set the features; the Internal BA introduces too much friction and doesn't allow for a clean rollback of prerequisites.

**Reply by barnson** [reply]
WixStandardBootstrapperApplication != WixInternalUIBootstrapperApplication. WixStandardBootstrapperApplication supports (for example) check boxes that set variables you can pass to MSI packages to enable/disable features.

**Reply by wmanning** [reply]
Yeah Internal BA is if you want the burn UI to be as invisible as possible and use the MSI UI. This is for people who want to use burn to boostrap a prerequisite (say .net) and then run their MSI UI for interaction. 

WixStandardBootstrapperApplication (wixstba for short) has some customization. 
- See the [WixStandardBootstrapperApplication](https://wixtoolset.org/docs/schema/bal/wixstandardbootstrapperapplication/#attributes) element.
- Doc intro to wixstba: https://wixtoolset.org/docs/tools/burn/wixstdba/
- There's a SuppressOptionsUI attribute that if you set to no (maybe that's default) you get an options page you can add controls to.  You could add elements there so users could select their desired features.
     - See my answer here: https://github.com/orgs/wixtoolset/discussions/8015
- You can also create your own theme file to give wixstba almost a custom look.  We do this to add our company logo, etc.
- My answer here may help as well: https://github.com/orgs/wixtoolset/discussions/7828

**Reply by amyspark** [reply]
> WixStandardBootstrapperApplication != WixInternalUIBootstrapperApplication. WixStandardBootstrapperApplication supports (for example) check boxes that set variables you can pass to MSI packages to enable/disable features.

I'm already aware that the two are different. However, according to the documentation, [setting `SuppressOptions=no` in the bootstrapper](https://wixtoolset.org/docs/schema/bal/wixstandardbootstrapperapplication/) and [`EnableFeatureSelection=yes` in the MsiPackage](https://wixtoolset.org/docs/schema/wxs/msipackage/) is supposed to be enough to cause the features to be listed in the Options dialog.

I have already searched the discussions and other Google results prior to posting this discussion, and could not find anything short of rolling out a custom BA. This process seems to be very poorly documented, and having to do that just to enable bundling the MSVC redistributable feels like a big drawback from our existing MSI-only installation process.

**Reply by barnson** [reply, marked as answer]
> cause the features to be listed in the Options dialog.

That's not the case. There is no stock BA that offers feature trees. _That_ does require a custom BA.

**Reply by wmanning** [reply]
> However, according to the documentation, [setting SuppressOptions=no in the bootstrapper](https://wixtoolset.org/docs/schema/bal/wixstandardbootstrapperapplication/) and [EnableFeatureSelection=yes in the MsiPackage](https://wixtoolset.org/docs/schema/wxs/msipackage/) is supposed to be enough to cause the features to be listed in the Options dialog.

Options just refers to an options page (where you can add custom controls) not features.
